File: 10-regular-expression-matching/regular-expression-matching.java
class Solution {
    public String sg;
    public String pg;

    public boolean isMatch(String s, String p) {
        sg = s;
        pg = p;
        return fun(0, 0);
    }

    public boolean fun(int sIndex, int pIndex) {
        if (sIndex == sg.length() && pIndex == pg.length()) {
            // System.out.println("5");
            return true;
        }

        if (pIndex == pg.length()) {
            // System.out.println("1");
            return false;
        }
        if (pIndex + 1 < pg.length() && pg.charAt(pIndex + 1) == '*') {
            // Case 1: skip 'x*'
            if (fun(sIndex, pIndex + 2)) {
                return true;
            }
            // Case 2: use '*' if current matches
            if (sIndex < sg.length() && (pg.charAt(pIndex) == sg.charAt(sIndex) || pg.charAt(pIndex) == '.')) {
                if (fun(sIndex + 1, pIndex)) {
                    return true;
                }
            }
            return false;
        }
        if (sIndex < sg.length() &&
                (pg.charAt(pIndex) == sg.charAt(sIndex) || pg.charAt(pIndex) == '.')) {
            if (fun(sIndex + 1, pIndex + 1)) {
                return true;
            }
        }

        // System.out.println("4");
        return false;
    }
}

File: 10-regular-expression-matching/regular-expression-matching.py
"""
p
aab


  p
c*a*b

"""

class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        def fun(p1,p2):
            if p2 == len(p):
                return p1 == len(s)
            if p1 == len(s):
                while p2 + 1 < len(p) and p[p2+1] == '*':
                    p2 += 2
                return p2 == len(p)
            if p2 + 1 < len(p) and p[p2+1] == '*':
                if p[p2] == '.':
                    if fun(p1,p2+2) or fun(p1+1,p2):
                        return True
                else:
                    i = p1
                    if fun(i,p2+2):
                        return True
                    while i < len(s) and s[i] == p[p2]:
                        if fun(i+1,p2+2):
                            return True
                        i+=1
            elif s[p1] == p[p2] or p[p2] == '.':
                if fun(p1+1,p2+1):
                    return True
            return False
        return fun(0,0)
